src/main/java/gregtech/api/unification/OreDictUnifier.java

```diff
@@ -165,20 +171,54 @@ public static void onItemRegistration(OreRegisterEvent event) {
         }
     }
 
-    public static Set<String> getOreDictionaryNames(ItemStack itemStack) {
+    @Nonnull
+    public static Set<String> getOreDictionaryNames(@Nonnull ItemStack itemStack) {
         if (itemStack.isEmpty()) return Collections.emptySet();
-        ItemAndMetadata key = new ItemAndMetadata(itemStack);
-        Set<String> names = stackOreDictName.get(key);
-        Set<String> wildcardNames = key.isWildcard() ? null : stackOreDictName.get(key.toWildcard());
+        ItemVariantMap<Set<String>> nameEntry = stackOreDictName.get(itemStack.getItem());
+        if (nameEntry == null) return Collections.emptySet();
+        short itemDamage = (short) GTUtility.getActualItemDamageFromStack(itemStack);
+        Set<String> names = nameEntry.getEntry(itemDamage);
+        Set<String> wildcardNames = itemDamage == GTValues.W ? null : nameEntry.getEntry(GTValues.W);
         if (names == null) {
             return wildcardNames == null ? Collections.emptySet() : Collections.unmodifiableSet(wildcardNames);
-        } else if (wildcardNames == null) {
+        } else if (wildcardNames == null || names == wildcardNames) { // single variant items have identical entries
             return Collections.unmodifiableSet(names);
         } else {
             return Sets.union(names, wildcardNames);
         }
     }
 
+    @Nullable
+    public static ItemVariantMap<Set<String>> getOreDictionaryEntry(@Nonnull Item item) {
+        ItemVariantMap.Mutable<Set<String>> entry = stackOreDictName.get(item);
+        return entry == null ? null : ItemVariantMap.unmodifiableSetView(entry);
+    }
+
+    @Nonnull
+    public static ItemVariantMap<Set<String>> getOreDictionaryEntryOrEmpty(@Nonnull Item item) {
+        ItemVariantMap.Mutable<Set<String>> entry = stackOreDictName.get(item);
+        return entry == null ? ItemVariantMap.empty() : ItemVariantMap.unmodifiableSetView(entry);
+    }
+
+    public static boolean hasOreDictionaryEntry(@Nonnull Item item) {
+        return stackOreDictName.containsKey(item);
+    }
+
+    public static boolean hasOreDictionary(@Nonnull ItemStack itemStack, @Nonnull String oreDictName) {
+        if (itemStack.isEmpty()) return false;
+        ItemVariantMap<Set<String>> nameEntry = stackOreDictName.get(itemStack.getItem());
+        if (nameEntry == null) return false;
+
+        short itemDamage = (short) GTUtility.getActualItemDamageFromStack(itemStack);
+        Set<String> names = nameEntry.getEntry(itemDamage);
+        if (names != null && names.contains(oreDictName)) return true;
+
+        if (itemDamage == GTValues.W) return false;
```

Comment:
Should this check be moved before retrieving the `names` set, for a faster early failure?

Reply:
The check exists to prevent itemstacks with wildcard damage value to fetch entry twice. Moving the check above will produce incorrect results.